server/src/main/java/org/opensearch/cluster/metadata/QueryGroup.java

```diff
@@ -0,0 +1,300 @@
+        }
+        validateResourceLimits(resourceLimits);
+        if (!isValid(updatedAt)) {
+            throw new IllegalArgumentException("QueryGroup.updatedAtInMillis is not a valid epoch");
+        }
+
+        this.name = name;
+        this._id = _id;
+        this.resiliencyMode = resiliencyMode;
+        this.resourceLimits = resourceLimits;
+        this.updatedAtInMillis = updatedAt;
+    }
+
+    private static boolean isValid(long updatedAt) {
+        long minValidTimestamp = Instant.ofEpochMilli(0L).getMillis();
+
+        // Use Instant.now() to get the current time in seconds since epoch
+        long currentSeconds = Instant.now().getMillis();
+
+        // Check if the timestamp is within a reasonable range
+        return minValidTimestamp <= updatedAt && updatedAt <= currentSeconds;
+    }
+
+    public QueryGroup(StreamInput in) throws IOException {
+        this(in.readString(), in.readString(), QueryGroupMode.fromName(in.readString()), in.readMap(), in.readLong());
+    }
+
+    /**
+     * Write this into the {@linkplain StreamOutput}.
+     *
+     * @param out
+     */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(name);
+        out.writeString(_id);
+        out.writeString(resiliencyMode.getName());
+        out.writeMap(resourceLimits);
+        out.writeLong(updatedAtInMillis);
+    }
+
+    private void validateResourceLimits(Map<String, Object> resourceLimits) {
+        for (Map.Entry<String, Object> resource : resourceLimits.entrySet()) {
+            String resourceName = resource.getKey();
+            Double threshold = (Double) resource.getValue();
+            Objects.requireNonNull(resourceName, "resourceName can't be null");
+            Objects.requireNonNull(threshold, "resource value can't be null");
+
+            if (Double.compare(threshold, 1.0) > 0) {
+                throw new IllegalArgumentException("resource value should be less than 1.0");
+            }
+
+            if (!ALLOWED_RESOURCES.contains(resourceName.toLowerCase(Locale.ROOT))) {
+                throw new IllegalArgumentException(
+                    "resource has to be valid, valid resources " + ALLOWED_RESOURCES.stream().reduce((x, e) -> x + ", " + e).get()
+                );
+            }
+        }
+    }
+
+    /**
+     * @param builder
+     * @param params
+     * @return
+     * @throws IOException
+     */
+    @Override
+    public XContentBuilder toXContent(final XContentBuilder builder, final Params params) throws IOException {
+        builder.startObject();
+//        builder.startObject(this._id);
+        builder.field("_id", _id);
+        builder.field("name", name);
+        builder.field("resiliency_mode", resiliencyMode.getName());
+        builder.field("updatedAt", updatedAtInMillis);
+        builder.mapContents(resourceLimits);
```

Comment:
no, this will flatten out the map into the `QueryGroup` object.
```
 /** writes a map without the start object and end object headers */
    public XContentBuilder mapContents(Map<String, ?> values) throws IOException {
        return map(values, true, false);
    }
```

Reply:
Yeah, I am wondering if we should keep the limits grouped instead of flattening them out. That keeps the schema more structured IMO

Reply:
Initially I was also thinking of the same but you suggested to keep it this way. But I agree that this is more structured.